src/signatures/secp256k1_ecdsa.rs

```diff
@@ -0,0 +1,210 @@
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct PublicKey(k256::ecdsa::VerifyingKey);
+
+impl PublicKey {
+    pub const LENGTH: usize = 33;
+
+    pub fn verify(&self, sig: &Signature, msg: &[u8]) -> bool {
+        use k256::ecdsa::signature::Verifier;
+        self.0.verify(msg, &sig.0).is_ok()
+    }
+
+    pub fn to_bytes(self) -> [u8; PublicKey::LENGTH] {
+        let encoded_point = self.0.to_encoded_point(true);
+        let slice = encoded_point.as_bytes();
+        let mut bytes = [0_u8; PublicKey::LENGTH];
+        bytes.copy_from_slice(slice);
+        bytes
+    }
+
+    pub fn try_from_bytes(bytes: &[u8; PublicKey::LENGTH]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed bytes",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    pub fn try_from_slice(bytes: &[u8]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed slice",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    // credit: [secret_key_to_address](https://github.com/gakonst/ethers-rs/)
+    pub fn to_address(&self) -> Address {
+        // let public_key = secret_key.verifying_key();
+        let public_key = self.0.to_encoded_point(/* compress = */ false);
+        let public_key = public_key.as_bytes();
+        debug_assert_eq!(public_key[0], 0x04);
+        // let hash = keccak256(&public_key[1..]);
+        use tiny_keccak::{Hasher, Keccak};
+        let mut keccak = Keccak::v256();
+        keccak.update(&public_key[1..]);
+        let mut hash = [0u8; 32];
+        keccak.finalize(&mut hash);
+
+        let mut bytes = [0u8; 20];
+        bytes.copy_from_slice(&hash[12..]);
+        Address::from(bytes)
+    }
+}
+
+impl TryFrom<[u8; PublicKey::LENGTH]> for PublicKey {
+    type Error = crate::Error;
+    fn try_from(bytes: [u8; PublicKey::LENGTH]) -> crate::Result<Self> {
+        Self::try_from_bytes(&bytes)
+    }
+}
+
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let encoded_point = self.0.to_encoded_point(true);
+        encoded_point.as_bytes().hash(state);
+    }
+}
+
+pub struct Signature(k256::ecdsa::Signature, k256::ecdsa::RecoveryId);
+
+impl Signature {
+    pub const LENGTH: usize = 65;
+
+    pub fn to_bytes(&self) -> [u8; Signature::LENGTH] {
+        let mut bytes = [0_u8; Signature::LENGTH];
+        bytes[0..64].copy_from_slice(&self.0.to_bytes());
+        bytes[64] = self.1.into();
+        bytes
+    }
+
+    pub fn try_from_bytes(sig: &[u8; Signature::LENGTH]) -> crate::Result<Self> {
+        let rid = k256::ecdsa::RecoveryId::from_byte(sig[64]).ok_or(crate::Error::ConvertError {
```

Comment:
`sig[64]` would panic if the bytes don't have the correct size

Reply:
Maybe you can just call `try_from_slice` since it checks the size

Reply:
`sig[64]` wouldn't panic because `sig: &[u8; Signature::LENGTH]` and `Signature::LENGTH == 65`. Maybe, one method would be enough. I don't like to constantly checking for slice size if its size is known at compile time. But taking fixed-size array from a slice is not very easy.

Reply:
Ah damn sorry, was reading both at the same time and I guess I got confused